File: server/router/rag.ts
import { z } from 'zod';
import { router, publicProcedure } from './trpc.js';
import {
  syncJournalEmbeddings,
  syncAllJournalEmbeddings,
  searchForContext,
  getEmbeddingStats,
} from '../ai/rag.js';
import { isEmbeddingsConfigured } from '../ai/embeddings.js';

export const ragRouter = router({
  // Check if embeddings are configured
  isConfigured: publicProcedure.query(async () => {
    return isEmbeddingsConfigured();
  }),

  // Get embedding stats
  getStats: publicProcedure.query(async () => {
    return getEmbeddingStats();
  }),

  // Sync a single journal entry
  syncJournal: publicProcedure
    .input(z.object({ journalId: z.string() }))
    .mutation(async ({ input }) => {
      const chunks = await syncJournalEmbeddings(input.journalId);
      return { chunks };
    }),

  // Sync all journal entries
  syncAll: publicProcedure.mutation(async () => {
    const result = await syncAllJournalEmbeddings();
    return result;
  }),

  // Semantic search
  search: publicProcedure
    .input(
      z.object({
        query: z.string(),
        limit: z.number().min(1).max(20).default(5),
      })
    )
    .query(async ({ input }) => {
      const results = await searchForContext(input.query, input.limit);
      return results;
    }),
});

File: backend/routes/rag.py
from flask import Blueprint, jsonify, request
from backend.ai.embeddings import is_embeddings_configured
from backend.ai.rag import (
    sync_journal_embeddings, sync_all_journal_embeddings,
    search_for_context, get_embedding_stats,
)

bp = Blueprint('rag', __name__, url_prefix='/api/rag')


@bp.get('/configured')
def configured():
    return jsonify(is_embeddings_configured())


@bp.get('/stats')
def stats():
    return jsonify(get_embedding_stats())


@bp.post('/sync/<journal_id>')
def sync_journal(journal_id):
    chunks = sync_journal_embeddings(journal_id)
    return jsonify({'chunks': chunks})


@bp.post('/sync-all')
def sync_all():
    result = sync_all_journal_embeddings()
    return jsonify(result)


@bp.get('/search')
def search():
    query = request.args.get('query', '').strip()
    limit = min(int(request.args.get('limit', 5)), 20)
    if not query:
        return jsonify([])
    return jsonify(search_for_context(query, limit))
